### PR TITLE
Standardize misc graph interface

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?

#4425 added syntactic sugar to CudaGraphManager to make it easy to graph arbitrary, non-layer, torch ops which may be keyed on things other than `padded_batch_dimensions`. This common sugar will be shared by several pieces of upcoming code.

In parallel, #4260 added similar syntactic sugar, but special-cased specifically to MTP CUDA Graphs. The implementation is very similar to that of #4425.

This PR subsumes MTP CUDA graphs into the same sugar as #4425, reconciling #4260's implementation with that of other planned work.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/models/common/language_module/language_module.py [hunk 1]
```diff
@@ -355,14 +357,16 @@ def compute_mtp_single_step(
...
             next_token_ids (Tensor): Correct next token IDs [1, N].
             position_ids (Tensor): Position IDs for the next tokens [1, N].
-            depth (int, optional): MTP depth index. Only needed when
-                ``mtp_use_repeated_layer`` is False (each depth uses a
-                distinct layer). Omit for repeated-layer models so that a
+            depth (int, optional): MTP depth index. Only needed when `mtp_use_repeated_layer` is
+                False (each depth uses a distinct layer). Omit for repeated-layer models so that a
                 single CUDA graph can serve all depths.
+            eager: Added as an optional argument here to simplify upstream code.
+            cache_key: Added as an optional argument here to simplify upstream code.
```
Comment (mathemakitten) [inline megatron/core/models/common/language_module/language_module.py]: These comments out-of-context are not useful to a reader (i.e. what upstream code?). So if we attempt to call MTP with a graph with a dimension which was not captured, this will currently fall through to re-capture?

File: megatron/core/models/common/language_module/language_module.py [hunk 2]
```diff
@@ -355,14 +357,16 @@ def compute_mtp_single_step(
...
 
         Returns:
             tuple: (new_hidden_states, logits [N, 1, vocab_size]).
         """
+        del eager, cache_key
```
Comment (mathemakitten) [inline megatron/core/models/common/language_module/language_module.py:372]:
```suggestion
        # CudaGraphManager consumes these args, if it exists 
        del eager, cache_key
```

File: megatron/core/inference/engines/dynamic_engine.py [hunk 3]
```diff
@@ -370,17 +370,17 @@ def create_cuda_graphs(self, reset_context: bool = True):
...
             tp_size = get_pg_size(controller.inference_wrapped_model.tp_group)
             sp_enabled = model_config.sequence_parallel and tp_size > 1
             mtp_pass_depth = not unwrapped.mtp.mtp_use_repeated_layer
             mtp_warmup_depths = range(controller._num_mtp_depths) if mtp_pass_depth else [None]
             mtp_seen_batch_sizes = set()
+        else:
+            tp_size = 1
+            sp_enabled = False
+            mtp_warmup_depths = None
+            mtp_seen_batch_sizes = None
```
Comment (mathemakitten) [inline megatron/core/inference/engines/dynamic_engine.py]: What is the point of setting these in the else block? It seems incorrect to do so? Right now if `sp_enabled = model_config.sequence_parallel and tp_size > 1` is True but we're not in MTP mode then `sp_enabled` will always be marked as False.

Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: The point was to avoid the linter from failing due to errors saying "variable may not have been defined before this point". That was all.

Let me fix this properly.

Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: Addressed.

File: megatron/core/models/common/language_module/language_module.py (hunk 1 above)
Comment (tdene) [reply megatron/core/models/common/language_module/language_module.py]: How's this?

**Review by mathemakitten — APPROVED** [review]

File: tests/unit_tests/transformer/test_cuda_graphs.py [hunk 4]
```diff
@@ -1126,11 +1144,18 @@ def teardown_method(self, method):
...
     @torch.inference_mode()
-    def test_inline_capture_matches_eager(self):
+    def test_inline_capture_matches_eager(self, make_module):
         """Inline-captured graph output must match eager execution."""
         config = self._make_config()
-        module = _SimpleModule(config).cuda().eval()
+        module = make_module(config)
 
         # Get eager reference before wrapping
         x = torch.randn(4, config.hidden_size, device="cuda")
```
Comment (santhnm2) [inline tests/unit_tests/transformer/test_cuda_graphs.py:1161]: Can we use the existing model building logic for this? Seems more rigorous than a dummy model.

**Review by santhnm2 — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Review by shanmugamr1992 — APPROVED** [review]
